DIPs/1NNN-AP.md

```diff
@@ -0,0 +1,172 @@
+| Status:         | Draft                                                           |
+
+## Abstract
+
+Enum Type Inference (ETI) is a shortcut to allow the omission of an enum member's
+type name when it can be contextually inferred.
+
+While D's pre-existing syntax makes ETI a more complex feature to implement than in brand
+new languages, its presence in many new languages has proven its convenience and popularity.
+
+## Contents
+* [Rationale](#rationale)
+* [Prior Work](#prior-work)
+* [Description](#description)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+Writing the same enum type names over and over again (e.g. exhaustive enum switch-cases)
+involves the unnecessary, often gratuitous repetition of the enum member's type name.
+
+The solution used by many other modern language is simple: permit the omission
+of the enum member's type name when it can be inferred from its context.
+Dlang having this shortcut would be equally beneficial, with few drawbacks.
+
+## Prior Work
+Implementation of this feature in other languages:
+- [Swift](https://docs.swift.org/swift-book/LanguageGuide/Enumerations.html)
+- [Ziglang](https://ziglang.org/documentation/master/#Enum-Literals)
+- [Odin](https://odin-lang.org/docs/overview/#implicit-selector-expression)
+
+Java allows omitting an enum type in switch-case statements: [https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html](https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html)
+
+A suggestion for a similar feature in Rust, but with a different approach: [https://internals.rust-lang.org/t/enum-path-inference-with-variant-syntax/16851](https://internals.rust-lang.org/t/enum-path-inference-with-variant-syntax/16851)
+
+## Description
+Due to `.enumMember` being the same syntax as the [module scope operator](https://dlang.org/spec/module.html#module_scope_operators),
+I propose that ETI should use the syntax `$enumMember`. I'm open to discussing alternatives,
+but I believe that my proposed syntax represents the best compromise between convenience and compatibility.
+
+#### 1. Assignment statements.
+Assigning to a variable with a known type should allow ETI.
+```d
+enum A{ a,b,c,d; }
+
+struct B{ A one, two; }
+
+void main(){
+    A    myA1 = $b; //myA1 = A.b
+    auto myA2 = $b; //error, we don't know the type of "$b"!
+    
+    B myB;
+    myB.one = $c; //myB.one = A.c
+    myB.two = $d; //myB.two = A.d
+}
+```
+
+#### 2. Return statements.
+If a function has an explicit enum return type, its return statement(s)
+should allow ETI.
+```d
+enum A{ a,b,c,d; }
+
+A myFunc(){
+    return $c; //returns A.c
+}
+
+auto myBrokenFunc(){
+    return $c; //error, we don't know the type of "$c"!
+}
+```
+
+#### 3. Parameters.
+Most parameters require explicit typing beforehand, and thus should
+always allow ETI. The exception is `alias` template parameters, which
```

Comment:
This change does not reflect the intended meaning of this sentence, which is referring to `alias` template parameters: `template(alias param);`
Template type parameters cannot be inferred from ETI enum members, as they do not contain a type. I showed this in the example below the suggested change. I didn't think it was worth mentioning specifically since I felt it was a given, whereas alias parameters could *theoretically* work with ETI if it allowed for incredibly ambiguous syntax.
Thanks for the suggestion though. :)

Reply:
The wording is *the exception*. So the example is surprising to the reader as they read that alias parameters are the only exception. BTW alias parameters can be typed `alias Foo foo`. There are also sequence parameters `seq...` which can take values of any type.

Reply:
```suggestion
always allow ETI. The exception is template parameters of unspecified type, which
```

Reply:
> The wording is _the exception_. So the example is surprising to the reader as they read that alias parameters are the only exception. BTW alias parameters can be typed `alias Foo foo`. There are also sequence parameters `seq...` which can take values of any type.

Okay, I think I'll just reword that bit entirely.